### ISSUE TITLE
Verify uploaded cookbook archives

### ISSUE THREAD

**Issue by alanivey** [chef/supermarket#692]
I came across this issue recently and @juliandunn suggested I post an issue as an Enhancement request.

The [system](https://supermarket.getchef.com/cookbooks/system) cookbook appears to have a problem with the tarball uploaded for v0.3.2. BSD tar is able to correctly handle it:

``` bash
$ tar -zxpvf system.tar.gz
x system/
x system/.gitignore
x system/.kitchen.yml
x system/.rubocop.yml
... snip ...
x system/test/unit/
x system/test/unit/spec/
x system/test/unit/spec/default_spec.rb
x system/test/unit/spec/spec_helper.rb
```

Not sure if related, but, the unpacked files have all permissions, as if the umask was 000 when creating; maybe it was created on Windows?

``` bash
$ cd system
$ ls -l
total 152
-rwxrwxrwx@  1 alan  staff     23 Apr 11 23:55 Berksfile
-rwxrwxrwx@  1 alan  staff    226 Apr 11 23:55 Berksfile.lock
-rwxrwxrwx@  1 alan  staff    782 Apr 11 23:55 CHANGELOG.md
-rwxrwxrwx@  1 alan  staff    248 Apr 11 23:55 Gemfile
-rwxrwxrwx@  1 alan  staff   5921 Apr 11 23:55 Gemfile.lock
-rwxrwxrwx@  1 alan  staff  10850 Apr 11 23:55 LICENSE
-rwxrwxrwx@  1 alan  staff   2220 Apr 11 23:55 README.md
-rwxrwxrwx@  1 alan  staff    289 Apr 11 23:55 Strainerfile
-rwxrwxrwx@  1 alan  staff    429 Apr 11 23:55 TESTING.md
-rwxrwxrwx@  1 alan  staff    241 Apr 11 23:55 Thorfile
-rwxrwxrwx@  1 alan  staff    492 Apr 11 23:55 Vagrantfile
drwxrwxrwx@  3 alan  staff    102 Apr 12 00:09 attributes
-rwxrwxrwx@  1 alan  staff    985 Apr 11 23:55 chefignore
drwxrwxrwx@  2 alan  staff     68 Apr 11 23:55 definitions
drwxrwxrwx@  3 alan  staff    102 Apr 12 00:09 files
drwxrwxrwx@  4 alan  staff    136 Apr 12 00:09 libraries
-rwxrwxrwx@  1 alan  staff   7640 Apr 11 23:55 metadata.json
-rwxrwxrwx@  1 alan  staff   4993 Apr 11 23:55 metadata.rb
drwxrwxrwx@  4 alan  staff    136 Apr 12 00:09 providers
drwxrwxrwx@ 10 alan  staff    340 Apr 12 00:09 recipes
drwxrwxrwx@  4 alan  staff    136 Apr 12 00:09 resources
drwxrwxrwx@  3 alan  staff    102 Apr 12 00:09 templates
drwxrwxrwx@  4 alan  staff    136 Apr 12 00:09 test
```

But, GNU tar pukes right away on the compression (I'm on a Mac, and `brew install coreutils` installs GNU tar as `gtar`):

``` bash
$ gtar -zxpvf system.tar.gz
gzip: unknown compression format
gtar: Child returned status 1
gtar: Error is not recoverable: exiting now
```

So, what does this have to do with Supermarket? If I use berkshelf to get this cookbook, it, too, pukes on the compression format:

``` bash
$ berks
Resolving cookbook dependencies...
Fetching cookbook index from https://supermarket.getchef.com...
Using apt (2.4.0)
... snip ...
Using selinux (0.8.0)
Installing system (0.3.2)
Using windows (1.32.3)
Using yum (3.2.2)
Using vim (1.1.2)
Using yum-epel (0.3.6)
E, [2014-07-30T12:05:10.046002 #5609] ERROR -- : Actor crashed!
Berkshelf::UnknownCompressionType: The file at '/var/folders/97/qtcflz1s6_q54cp2blsl06n80000gn/T/community-rest-stream20140730-5609-qwszy9' is not a known compression type
    /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/community_rest.rb:19:in `unpack'
... snip ...
```

I'm not sure if Berkshelf is using GNU tar or GNU gzip to uncompress the archive, but it is failing for the same reason that my manual GNU tar command does.

My request: Supermarket should verify an uploaded cookbook archive can, in fact, be uncompressed and unpacked. If it already does, it should be done in the same manner and/or with the same libraries as Berkshelf.

I realize this may be an edge-case. I'll contact the cookbook author and request that they re-upload the cookbook archive. But, depending on how this archive was created, it very may well could come up again, and perhaps it would be a good idea for Supermarket to reject the archive if it cannot be fully unpacked using GNU libraries.

**TL;DR:** Uploaded cookbook archive cannot be unpacked because of an incompatibility with the archive file and either GNU tar or GNU gzip. Request that Supermarket perform [a different?] validation on the uploaded file in the same manner that Berkshelf uses to unpack.
